Migrate Sympy Lecture from the Intro Series (#284)
QuantEcon/lecture-python-programming: lectures/sympy.md

We can expand this expression with the `expand` function

```{code-cell} ipython3
expand_expr = expand(expr)
expand_expr
```

+++ {"user_expressions": []}

and factorize it back to the factored form with the `factor` function

```{code-cell} ipython3
factor(expand_expr)
```

+++ {"user_expressions": []}

We can solve this expression

```{code-cell} ipython3
solve(expr)
```

+++ {"user_expressions": []}

Note this is equivalent to solving the equation

$$
(x + y)^2 = 0 
$$

Another way to solve this equation is to use `solveset`.

`solveset` allows users to define the domain of `x` in the evaluation and output a set as a solution

```{code-cell} ipython3
solveset(expr, x, domain=S.Naturals)
```

```{note}
[Solvers](https://docs.sympy.org/latest/modules/solvers/index.html) is an important module with tools to solve different types of equations. 

There are a variety of solvers available in SymPy depending on the nature of the problem.
```

+++ {"user_expressions": []}

### Equations

SymPy provides several functions to manipulate equations.

Let's develop an equation with the expression we defined before

```{code-cell} ipython3
eq = Eq(expr, 0)
eq
```

+++ {"user_expressions": []}

> Solving this expression with respect to x gives the same output as solving the expression directly

Reviewer: ```suggestion
Solving this equation with respect to x gives the same output as solving the expression directly
```
?

$$
k^*:=\left(\frac{s A}{\delta}\right)^{1 /(1-\alpha)}
$$ 

This can be easily computed in SymPy

```{code-cell} ipython3
A, s, k, α, δ = symbols('A s k α δ')
```

```{code-cell} ipython3
# Define Solow-Swan growth dynamics
solow = Eq(s*A*k**α + (1 - δ) * k, k)
solow
```

```{code-cell} ipython3
solve(solow, k)
```

### Inequalities and logic

SymPy also allows users to define inequalities and set operators and provides a wide range of [operations](https://docs.sympy.org/latest/modules/solvers/inequalities.html).

```{code-cell} ipython3
reduce_inequalities([2*x + 5*y <= 30, 4*x + 2*y <= 20], [x])
```

```{code-cell} ipython3
And(2*x + 5*y <= 30, x > 0)
```

+++ {"user_expressions": []}

### Series

Series are widely used in economics and statistics, from asset pricing to the expectation of discrete random variables.

We can construct a simple series of summations using `Sum` function

```{code-cell} ipython3
x, y, i, j = symbols("x y i j")
sum_xy = Sum(Indexed('x',i) * Indexed('y', j), 
            (i, 0, 3),
            (j, 0, 3))
sum_xy
```

+++ {"user_expressions": []}

To evaluate the sum, we can `lamdify` the formula.

> The lamdified expression can take vectors as input for $x$ and $y$ and compute the result

Reviewer: ```suggestion
The lamdified expression can take numeric values as input for $x$ and $y$ and compute the result
```

Lambdify in general works for all numeric inputs. See: https://docs.sympy.org/latest/modules/utilities/lambdify.html#sympy.utilities.lambdify.lambdify

```{code-cell} ipython3
Dt.doit()
```

+++ {"user_expressions": []}

Simplifying the expression above gives

```{code-cell} ipython3
simplify(Dt.doit())
```

+++ {"user_expressions": []}

This is consistent with the solution we provided in our lecture on [geometric series](https://python.quantecon.org/geom_series.html#example-the-money-multiplier-in-fractional-reserve-banking).

#### Example: discrete random variable

We can also use SymPy to compute the expectation of a discrete random variable.

Let's define a discrete random variable $X$ with [Poisson distribution](https://en.wikipedia.org/wiki/Poisson_distribution):

$$
f(x) = \frac{\lambda^x e^{-\lambda}}{x!}, \quad x = 0, 1, 2, \ldots
$$

```{code-cell} ipython3
λ = symbols('lambda')

# We refine the symbol x to positive integers
x = Symbol('x', integer=True, positive=True)
pmf = λ**x * exp(-λ)/factorial(x)
pmf
```

We can verify if the distribution follows the fundamental property of probability distribution:

$$
\sum_{x=0}^{\infty} f(x) = 1
$$

```{code-cell} ipython3
sum_pmf = Sum(pmf, (x, 0, oo))
sum_pmf.doit()
```

The expectation of the distribution is:

$$
E(X) = \sum_{x=0}^{\infty} x f(x) 

> $$

Reviewer: This is nice. We can even introduce the symbolic stats using https://docs.sympy.org/latest/modules/stats.html#sympy.stats.Poisson. I've worked on the `stats` module in `sympy` :)

```{code-cell} ipython3
fx = Sum(x*pmf, (x, 0, oo))
fx.doit()
```

```{note}

SymPy offers a statistics module called [`Stats`](https://docs.sympy.org/latest/modules/stats.html).

`Stats` module offers built-in distributions and functions on probability distributions.

The operation above can also be computed using `Stats` module.
```

+++ {"user_expressions": []}

## Symbolic Calculus

SymPy allows us to perform various calculus operations, such as differentiation and integration.

### Limits

We can compute limits for a given function using the `limit` function

```{code-cell} ipython3
---
mystnb:
  image:
    width: 3%
---
# Define a function

> f = x ** 2 / (x - 1)

Reviewer: This is actually an expression and not a function in sympy. Function `F` would be:
```
F = Lambda(x, f)
```

  image:
    width: 65%
---
# Solve for phi
solution = solve(indifference_eq, phi)

# Simplify the solution
solution = simplify(solution[0])
solution
```

+++ {"user_expressions": []}

To compute a numerical value for $\phi$, we replace symbols $w$, $R$, $g$, and $D$ with specific numbers. 

Here we use $w = 1$, $R = 1.05$, $g = 0.02$, and $D = 0.5$. 

Substituting them into the equation, we find a specific value for $\phi$

```{code-cell} ipython3
---
mystnb:
  image:
    width: 20%
---
# Substitute specific values
solution_num = solution.subs({w: 1, R: 1.05, g: 0.02, D: 0.5})
solution_num
```

+++ {"user_expressions": []}

The wage of a college graduate is approximately 0.797 times the wage of a high school graduate.

+++ {"user_expressions": []}

```{exercise}
:label: sympy_ex1

L'Hôpital's rule states that for two functions $f(x)$ and $g(x)$, if $\lim_{x \to a} f(x) = \lim_{x \to a} g(x) = 0$ or $\pm \infty$, then

$$
\lim_{x \to a} \frac{f(x)}{g(x)} = \lim_{x \to a} \frac{f'(x)}{g'(x)}
$$

Use SymPy to verify L'Hôpital's rule for the following functions

$$
f(x) = \frac{y^x - 1}{x}
$$

as $x$ approaches to $0$
```

```{solution-start} sympy_ex1
:class: dropdown
```

+++

> Let's define the function first

Reviewer: Again, `f` here is an expression.

```{code-cell} ipython3
f_upper = y**x - 1
f_lower = x
f = f_upper/f_lower
f
```

Sympy is smart enough to solve this limit

```{code-cell} ipython3
lim = limit(f, x, 0)
lim
```

We compare the result suggested by L'Hôpital's rule

```{code-cell} ipython3
lim = limit(diff(f_upper, x)/
            diff(f_lower, x), x, 0)
lim
```

```{solution-end}
```

```{exercise}
:label: sympy_ex2

Maximum likelihood estimation (MLE) is a method to estimate the parameters of a statistical model. 

It usually involves maximizing a log-likelihood function and solving the first-order derivative.

In this exercise, we use SymPy to compute the MLE of a binomial distribution.
```

```{solution-start} sympy_ex2
:class: dropdown
```

> First, we define the binomial distribution

Reviewer: Sympy provides this too: https://docs.sympy.org/latest/modules/stats.html#sympy.stats.Binomial

Author: Many thanks @Smit-create for your review,

I acknowledged that SymPy has a module on statistics in a note directive, but I did not include it because the lecture gets super long now : )

Maybe I should include a sentence or two to run through a quick example in conjunction with the implementations from scratch.

---
jupytext:
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.14.5
kernelspec:
  display_name: Python 3 (ipykernel)
  language: python
  name: python3
---

> 

Reviewer 2: There doesn't appear to be a title to this lecture?

```{code-cell} ipython3

> !pip install sympy==1.12

Reviewer 3: ```suggestion
---
tags: [hide-output]
---

!pip install sympy==1.12
```

Reviewer 2: Is there a reason we need a specific version of sympy? Do updates often break and cause instability?

Author: Hi @mmcky,

This line was previously used to test whether [a feature in SymPy](https://github.com/sympy/sympy/pull/24660) has been included in the latest version of SymPy. The `pip` install has been removed because it is included in the base anaconda.

```

+++ {"user_expressions": []}

# SymPy

## Overview

Unlike numerical libraries that deal with values, [SymPy](https://www.sympy.org/en/index.html) focuses on manipulating mathematical symbols and expressions directly.

SymPy provides [a wide range of functionalities](https://www.sympy.org/en/features.html) including 

- Symbolic Expression
- Equation Solving
- Simplification
- Calculus
- Matrices
- Discrete Math, etc.

These functions make SymPy a popular open-source alternative to other proprietary symbolic computation software such as Mathematica.

In this lecture, we will explore some of the functionalities of SymPy and demonstrate how to use basic SymPy functions to solve economic models.

## Getting Started

Let’s first import the library and initialize the printer for symbolic output

```{code-cell} ipython3

> from sympy import *

Reviewer 2: Is this the recommended way to work with `sympy`? It is often a big no no in Python land -- but there are probably a LOT of imports needed in `sympy` case for things like `cos` and `tan` etc.

Author: Hi @mmcky,

This lecture includes a lot of SymPy functions, so I break the rule here. This is also what they have done in the [SymPy tutorial](https://docs.sympy.org/latest/tutorials/intro-tutorial/basic_operations.html).

from sympy.plotting import plot, plot3d_parametric_line, plot3d
from sympy.solvers.inequalities import reduce_rational_inequalities
from sympy.stats import Poisson, Exponential, Binomial, density, moment, E

import numpy as np

# Enable the best printer available
init_printing()
```

+++ {"user_expressions": []}

## Symbolic algebra

### Symbols

+++ {"user_expressions": []}

Before we start manipulating the symbols, we initialize some symbols to work with

```{code-cell} ipython3
x, y, z = symbols('x y z')
```

+++ {"user_expressions": []}

Symbols are the basic units for symbolic computation in SymPy.

> +++ {"user_expressions": []}

Reviewer 2: These artifacts of `jupytext` should be deleted

### Expressions

We can now use symbols `x`, `y`, and `z` to build expressions and equations.

Here we build a simple expression first

```{code-cell} ipython3
---
mystnb:
  image:

>     width: 10%

Reviewer 2: @HumphreyYang @HengchengZhang should we add some rules to our manual around using symbolics?

Author: Hi @mmcky,

I just removed those headers, and the formula looks fine on my local build. @HengchengZhang, were these headers added because of the size of the formula?

```

+++ {"user_expressions": []}

### Equations

SymPy provides several functions to manipulate equations.

Let's develop an equation with the expression we defined before

```{code-cell} ipython3
eq = Eq(expr, 0)
eq
```

+++ {"user_expressions": []}

Solving this equation with respect to x gives the same output as solving the expression directly

```{code-cell} ipython3
solve(eq, x)
```

+++ {"user_expressions": []}

SymPy can easily handle equations with multiple solutions

```{code-cell} ipython3
eq = Eq(expr, 1)
solve(eq, x)
```

+++ {"user_expressions": []}

`solve` function can also combine multiple equations together and solve a system of equations

```{code-cell} ipython3
eq2 = Eq(x, y)
eq2
```

```{code-cell} ipython3
solve([eq, eq2], [x, y])
```

+++ {"user_expressions": []}

We can also solve for the value of $y$ by simply substituting $x$ with $y$

```{code-cell} ipython3
expr_sub = expr.subs(x, y)
expr_sub
```

```{code-cell} ipython3
solve(Eq(expr_sub, 1))
```

+++ {"user_expressions": []}

Below is another example equation with symbol `x` and functions `sin`, `cos`, and `tan` using the `Eq` function

```{code-cell} ipython3
---
mystnb:
  image:

>     width: 17.5%

Reviewer 2: There seems to be a lot of customisation to these sizes in `jb`

Author: I think @HengchengZhang made these changes to make the formula look better in the build. I was thinking about applying these configurations across the lecture.

Reviewer 3: Hi @mmcky and @HumphreyYang, I think sympy doesn’t really have size config for outputs so I add these to make them look better.

---
# Create an equation
eq = Eq(cos(x) / (tan(x)/sin(x)), 0)
eq
```

+++ {"user_expressions": []}

Now we simplify this equation using the `simplify` function

```{code-cell} ipython3
---
mystnb:
  image:

>     width: 10%

Reviewer 3: ```suggestion
    width: 11%
```

---
# Simplify an expression
simplified_expr = simplify(eq)
simplified_expr
```

+++ {"user_expressions": []}

We can solve equations using the `solve` function in SymPy

```{code-cell} ipython3
---
mystnb:
  image:

>     width: 5.5%

Reviewer 3: ```suggestion
    width: 7.5%
```

```{code-cell} ipython3
reduce_inequalities([2*x + 5*y <= 30, 4*x + 2*y <= 20], [x])
```

```{code-cell} ipython3
And(2*x + 5*y <= 30, x > 0)
```

+++ {"user_expressions": []}

### Series

Series are widely used in economics and statistics, from asset pricing to the expectation of discrete random variables.

We can construct a simple series of summations using `Sum` function

```{code-cell} ipython3
x, y, i, j = symbols("x y i j")
sum_xy = Sum(Indexed('x',i) * Indexed('y', j), 
            (i, 0, 3),
            (j, 0, 3))
sum_xy
```

+++ {"user_expressions": []}

To evaluate the sum, we can [`lamdify`](https://docs.sympy.org/latest/modules/utilities/lambdify.html#sympy.utilities.lambdify.lambdify) the formula.

The lamdified expression can take numeric values as input for $x$ and $y$ and compute the result

```{code-cell} ipython3
sum_xy = lambdify([x, y], sum_xy)
grid = np.arange(0, 4, 1)
sum_xy(grid, grid)
```

+++ {"user_expressions": []}

#### Example: bank deposits

Imagine a bank with $D_0$ as the deposit at time $t$, it loans $(1-r)$ of its deposits and keeps a fraction $r$ as cash reserves.

> Its deposits at time is

Reviewer 2: time? Should this be time `0` or something like that?

Author: Many thanks @mmcky, corrected.

pmf = λ**x * exp(-λ)/factorial(x)
pmf
```

+++ {"user_expressions": []}

We can verify if the distribution follows the fundamental property of probability distribution:

$$
\sum_{x=0}^{\infty} f(x) = 1
$$

```{code-cell} ipython3
sum_pmf = Sum(pmf, (x, 0, oo))
sum_pmf.doit()
```

+++ {"user_expressions": []}

The expectation of the distribution is:

$$
E(X) = \sum_{x=0}^{\infty} x f(x) 
$$

```{code-cell} ipython3
fx = Sum(x*pmf, (x, 0, oo))
fx.doit()
```

+++ {"user_expressions": []}

SymPy offers a statistics module called [`Stats`](https://docs.sympy.org/latest/modules/stats.html).

`Stats` module offers built-in distributions and functions on probability distributions.

The computation above can also be condensed into one line using `Stats` module

```{code-cell} ipython3
# Using sympy.stats.Poisson() method
λ = Symbol("λ", positive = True)
X = Poisson("x", λ)
E(X)
```

+++ {"user_expressions": []}

## Symbolic Calculus

SymPy allows us to perform various calculus operations, such as differentiation and integration.

### Limits

We can compute limits for a given expression using the `limit` function

```{code-cell} ipython3
---
mystnb:
  image:

>     width: 3%

Reviewer 3: ```suggestion
    width: 1.5%
```

Reviewer 2: What happens if we leave this metadata scaling out? Does Sympy do a bad job of fitting the equations into the correct size in `jupyter book`

E(X**t)
```

+++ {"user_expressions": []}

Using `integrate` function, we can derive the cumulative density function of the exponential distribution with $\lambda = 0.5$

```{code-cell} ipython3
lamda_pdf = pdf.subs(λ, 1/2)
lamda_pdf
```

```{code-cell} ipython3
integrate(lamda_pdf, (x, 0, 4))
```

+++ {"user_expressions": []}

## Plotting

SymPy provides a powerful plotting feature. 

First we plot a simple function using the `plot` function

```{code-cell} ipython3
f = sin(2 * sin(2 * sin(2 * sin(x))))
p = plot(f, (x, -10, 10), show=False)
p.title = 'A Simple Plot'
p.show()
```

+++ {"user_expressions": []}

Similar to Matplotlib, SymPy provides an interface to customize the graph

```{code-cell} ipython3
plot_f = plot(f, (x, -10, 10), 
              xlabel='', ylabel='', legend = True, show = False)
plot_f[0].label = 'f(x)'
df = diff(f)
plot_df = plot(df, (x, -10, 10), legend = True, show = False)
plot_df[0].label = 'f\'(x)'
plot_f.append(plot_df[0])
plot_f.show()
```

+++ {"user_expressions": []}

It also supports plotting implicit functions and visualizing inequalities

```{code-cell} ipython3
p = plot_implicit(Eq((1/x + 1/y)**2, 1))
```

```{code-cell} ipython3

> p = plot_implicit(And(2*x + 5*y <= 30, 4*x + 2*y >= 20), (x, -1, 10), (y, -10, 10))

Reviewer 3: ```suggestion
p = plot_implicit(And(2*x + 5*y <= 30, 4*x + 2*y >= 20),
                     (x, -1, 10), (y, -10, 10))
```

```

+++ {"user_expressions": []}

It also supports visualizations in three-dimensional space

```{code-cell} ipython3
p = plot3d(cos(2*x + y))
```

+++ {"user_expressions": []}

## Applications

In this section, we apply SymPy to an economic model that explores the wage gap between college and high school graduates. 

### Defining the Symbols

The first step in symbolic computation is to define the symbols that represent the variables in our equations. In our model, we have the following variables:

 * $R$ The gross rate of return on a one-period bond

 * $t = 0, 1, 2, \ldots T$ Denote the years that a person either works or attends college
 
 * $T$ denotes the last period that a person  works
 
 * $w$ The wage of a high school graduate
 
 * $g$ The growth rate of wages

 * $D$ The upfront monetary costs of going to college
 
 * $\phi$ The wage gap, defined as the ratio of the wage of a college graduate to the wage of a high school graduate
 
Let's define these symbols using sympy

```{code-cell} ipython3
# Define the symbols
w, R, g, D, phi = symbols('w R g D phi')
```

+++ {"user_expressions": []}

### Defining the Present Value Equations

> We calculate the present value of the earnings of a high school graduate and a college graduate. For a high school graduate, we sum the discounted wages from year 1 to $T = 4$. For a college graduate, we sum the discounted wages from year 5 to $T = 4$, considering the cost of college $D$ paid upfront

Reviewer 2: It can be useful to add a plot that shows the various flows `+/-` along a timeline. Would that be possible?

---
jupytext:
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.14.4
kernelspec:
  display_name: Python 3 (ipykernel)
  language: python
  name: python3
---

(sympy)=
```{raw} jupyter
<div id="qe-notebook-header" align="right" style="text-align:right;">
        <a href="https://quantecon.org/" title="quantecon.org">
                <img style="width:250px;display:inline;" width="250px" src="https://assets.quantecon.org/img/qe-menubar-logo.svg" alt="QuantEcon">
        </a>
</div>
```

# {index}`SymPy <single: SymPy>`

```{index} single: Python; SymPy
```

```{contents} Contents
:depth: 2
```

## Overview

Unlike numerical libraries that deal with values, [SymPy](https://www.sympy.org/en/index.html) focuses on manipulating mathematical symbols and expressions directly.

SymPy provides [a wide range of features](https://www.sympy.org/en/features.html) including 

> - Symbolic Expression

Reviewer 4: please change to lower case, in this bullet point and those below

- Equation Solving
- Simplification
- Calculus
- Matrices
- Discrete Math, etc.

> These functions make SymPy a popular open-source alternative to other proprietary symbolic computation software such as Mathematica.

Reviewer 4: "computational software"

> In this lecture, we will explore some of the functionalities of SymPy and demonstrate how to use basic SymPy functions to solve economic models.

Reviewer 4: functionalities -> functionality

## Getting Started

Let’s first import the library and initialize the printer for symbolic output

```{code-cell} ipython3
from sympy import *
from sympy.plotting import plot, plot3d_parametric_line, plot3d
from sympy.solvers.inequalities import reduce_rational_inequalities
from sympy.stats import Poisson, Exponential, Binomial, density, moment, E, cdf

import numpy as np
import matplotlib.pyplot as plt

# Enable the mathjax printer
init_printing(use_latex='mathjax')
```

## Symbolic algebra

### Symbols

> Before we start manipulating the symbols, we initialize some symbols to work with

Reviewer 4: "Before... symbols,"  -> "First"

```{code-cell} ipython3
x, y, z = symbols('x y z')
```

Symbols are the basic units for symbolic computation in SymPy.

### Expressions

We can now use symbols `x`, `y`, and `z` to build expressions and equations.

Here we build a simple expression first

```{code-cell} ipython3
expr = (x+y) ** 2
expr
```

We can expand this expression with the `expand` function

```{code-cell} ipython3
expand_expr = expand(expr)
expand_expr
```

and factorize it back to the factored form with the `factor` function

```{code-cell} ipython3
factor(expand_expr)
```

We can solve this expression

```{code-cell} ipython3
solve(expr)
```

Note this is equivalent to solving the following equation for `x`

$$
(x + y)^2 = 0 
$$

> 

Reviewer 4: I recommend that we cut from here...

Another way to solve this equation is to use `solveset`.

`solveset` allows users to define the domain of `x` in the evaluation and output a set as a solution

```{code-cell} ipython3
solveset(expr, x, domain=S.Naturals)

> ```

Reviewer 4: to here in the interest of brevity --- we don't have to mention all possibilities.

```{note}
[Solvers](https://docs.sympy.org/latest/modules/solvers/index.html) is an important module with tools to solve different types of equations. 

> 

Reviewer 4: module -> submodule?

# Create an equation
eq = Eq(cos(x) / (tan(x)/sin(x)), 0)
eq
```

Now we simplify this equation using the `simplify` function

```{code-cell} ipython3
# Simplify an expression
simplified_expr = simplify(eq)
simplified_expr
```

Again, we use the `solve` function to solve this equation

```{code-cell} ipython3
# Solve the equation
sol = solve(eq, x)
sol
```

SymPy can also handle more complex equations involving trigonometry and complex numbers.

We demonstrate this using [Euler's formula](https://en.wikipedia.org/wiki/Euler%27s_formula)

```{code-cell} ipython3
# 'I' represents the imaginary number i 
euler = cos(x) + I*sin(x)
euler
```

```{code-cell} ipython3
simplify(euler)
```

If you are interested, we encourage you to read the lecture on [trigonometry and complex numbers](https://python.quantecon.org/complex_and_trig.html).

#### Example: fixed point computation

Fixed point computation is frequently used in economics and finance.

Here we solve the fixed point of the Solow-Swan growth dynamics:

$$
k_{t+1}=s f\left(k_t\right)+(1-\delta) k_t, \quad t=0,1, \ldots
$$

where $k_t$ is the capital stock, $f$ is a production function, $\delta$ is a rate of depreciation.

> 

Reviewer 4: Perhaps it's a good idea to write down the equation being solved (for people who aren't completely sure what a fixed point is.

And(2*x + 5*y <= 30, x > 0)
```

### Series

Series are widely used in economics and statistics, from asset pricing to the expectation of discrete random variables.

We can construct a simple series of summations using `Sum` function and `Indexed` symbols

```{code-cell} ipython3
x, y, i, j = symbols("x y i j")
sum_xy = Sum(Indexed('x', i)*Indexed('y', j), 
            (i, 0, 3),
            (j, 0, 3))
sum_xy
```

To evaluate the sum, we can [`lambdify`](https://docs.sympy.org/latest/modules/utilities/lambdify.html#sympy.utilities.lambdify.lambdify) the formula.

The lambdified expression can take numeric values as input for $x$ and $y$ and compute the result

```{code-cell} ipython3
sum_xy = lambdify([x, y], sum_xy)
grid = np.arange(0, 4, 1)
sum_xy(grid, grid)
```

#### Example: bank deposits

Imagine a bank with $D_0$ as the deposit at time $t$.

It loans $(1-r)$ of its deposits and keeps a fraction $r$ as cash reserves.

Its deposits over an infinite time horizon can be written as

$$
\sum_{i=0}^\infty (1-r)^i D_0
$$

Let's compute the deposits at time $t$

```{code-cell} ipython3
D = symbols('D_0')
r = Symbol('r', positive=True)
Dt = Sum('(1 - r)^i * D_0', (i, 0, oo))
Dt
```

> We can call `doit` method to evaluate the series

Reviewer 4: "call" -> "call the"

Its deposits over an infinite time horizon can be written as

$$
\sum_{i=0}^\infty (1-r)^i D_0
$$

Let's compute the deposits at time $t$

```{code-cell} ipython3
D = symbols('D_0')
r = Symbol('r', positive=True)
Dt = Sum('(1 - r)^i * D_0', (i, 0, oo))
Dt
```

We can call `doit` method to evaluate the series

```{code-cell} ipython3
Dt.doit()
```

Simplifying the expression above gives

```{code-cell} ipython3
simplify(Dt.doit())
```

This is consistent with the solution in the lecture on [geometric series](https://python.quantecon.org/geom_series.html#example-the-money-multiplier-in-fractional-reserve-banking).

#### Example: discrete random variable

In the following example, we compute the expectation of a discrete random variable.

Let's define a discrete random variable $X$ following a [Poisson distribution](https://en.wikipedia.org/wiki/Poisson_distribution):

$$
f(x) = \frac{\lambda^x e^{-\lambda}}{x!}, \quad x = 0, 1, 2, \ldots
$$

```{code-cell} ipython3
λ = symbols('lambda')

# We refine the symbol x to positive integers
x = Symbol('x', integer=True, positive=True)
pmf = λ**x * exp(-λ) / factorial(x)
pmf
```

> We can verify if the distribution follows the fundamental property of probability distribution:

Reviewer 2: @HumphreyYang @chappiewuzefan I am not sure what this means? What is the fundamental property?

lambda_indiff
```

We use the following parameters to compute the value for $\phi$

```{code-cell} ipython3
R_value = 1.05
γ_c_value, γ_h_value = 1.01, 1.01
wh0_value = 1
wc0_value = 2
D_value = 10
T_value = 40

lambda_indiff(R_value, wh0_value, wc0_value, 
              γ_c_value, γ_h_value, D_value, 
              T_value)
```

Under this setting, we find college is more attractive to the person as $\phi < 1$.

Note that we can also use the `subs` method to substitute the values of the parameters

```{code-cell} ipython3
indiff_sol.subs({R: R_value, 
                wh0: wh0_value, 
                wc0: wc0_value, 
                γ_c: γ_c_value, 
                γ_h: γ_h_value, 
                D: D_value, 
                T: T_value})
```

It is also possible to use the [`evalf`](https://docs.sympy.org/latest/modules/evalf.html) method to get floating-point approximations

```{code-cell} ipython3
indiff_sol.evalf(subs={R: R_value, 
                      wh0: wh0_value, 
                      wc0: wc0_value, 
                      γ_c: γ_c_value, 
                      γ_h: γ_h_value, 
                      D: D_value, 
                      T: T_value})
```

> Note how precision changes using different methods.

Reviewer 2: Perhaps this could be:

    It is useful to be aware that precision can be different depending on the method used to evaluate a function or solve an expression. 

    We encourage readers to [read more here ]() 

With a link to the SymPy library. Is there a discussion on precision there we can point to?

Author: Many thanks @mmcky.

I think I will remove the second sentence as the documentation for `evalf` is pretty clear about the nature of floating point approximation.

                                γ_h: γ_h_value, 
                                D: D_value})
```

```{code-cell} ipython3
plt.plot(np.arange(10, 60, 1),
         lambdify(T, indiff_sol_T)(np.arange(10, 60, 1)))
plt.xlabel('T')
plt.ylabel(r'$\phi$')
plt.show()
```

As $T$ increases, $\phi$ decreases favoring attending college.

```{code-cell} ipython3
indiff_sol_TD = indiff_sol.subs({R: R_value, 
                                wh0: wh0_value, 
                                wc0: wc0_value, 
                                γ_c: γ_c_value, 
                                γ_h: γ_h_value})
```

What if we want to know how $\phi$ changes with different tuition fees $D$ and $T$?

```{code-cell} ipython3
grid = np.meshgrid(np.arange(10, 60, 1), 
                   np.arange(0, 60, 1))
```

```{code-cell} ipython3
ϕ_TR = lambdify([T, D], indiff_sol_TD)(grid[0], grid[1])
```

```{code-cell} ipython3
fig = plt.figure()
ax = plt.axes(projection ='3d')
ax.set_box_aspect(aspect=None, zoom=0.85)

ax.plot_surface(grid[0], 
                grid[1],
                ϕ_TR)
ax.set_xlabel('T')
ax.set_ylabel('D')
ax.set_zlabel(r'$\phi$')
plt.show()
```

> We can see how $\phi$ flats out when we increase the time horizon as college graduates have higher salary at different levels of tuition fee.

Reviewer 2: ```suggestion
We can see how $\phi$ flattens out when we increase the time horizon as college graduates have higher salary at different levels of tuition fee.
```
